bin/analyzeShearTest.py

```diff
@@ -0,0 +1,265 @@
+    #   signal-to-noise of some other field such as flux and flux error.  In that
+    #   case, the weighting value used is shape_field/(sigma_field/sigma_signal_field)
+    if "shape_field" in params.keys():
+        shape_field = params["shape_field"] + "_"
+        shape_type = params["shape_type"]
+        if "sigma_field" in params.keys():
+            sigma_field = params["sigma_field"]
+            if "sigma_signal_field" in params.keys():
+                sigma_signal_field = params["sigma_signal_field"]
+            else:
+                sigma_signal_field = None
+        else:
+            sigma_field = None
+            sigma_signal_field = None
+    else:
+        #   If shape_field is not in params, the ellipticity and sigma field are here:
+        shape_field = ""
+        shape_type = "ellipticity"
+        sigma_field = "e_sigma"
+        sigma_signal_field = None
+
+    #   Use the butler to get the src.fits file for each subfield/epoch
+    butler = dafPersist.Butler(root=os.path.join(args.base_dir, output_dir))
+
+    count = 0
+    eSum = 0.0
+    e1DevSum = 0.0
+    e2DevSum = 0.0
+    eSumSq = 0.0
+    e1DevSumSq = 0.0
+    e2DevSumSq = 0.0
+    thetaSum = 0.0
+    for subfield in range(n_subfields):
+        for epoch in range(n_epochs):
+            catCount = 0
+            catE1DevSum = 0.0
+            catE2DevSum = 0.0
+            catWeightSum = 0.0
+            sourceCat = butler.get("src", {'subfield':subfield, 'epoch':epoch})
+
+            #  These are constant values per subfield recorded by galsim
+            g1Key = sourceCat.getSchema().find("g1").getKey()
+            g2Key = sourceCat.getSchema().find("g2").getKey()
+
+            #   Set up the measurement keys, depending on type (moments or ellip)
+            if shape_type == "ellipticity":
+                e1Key = sourceCat.getSchema().find(shape_field + "e1").getKey()
+                e2Key = sourceCat.getSchema().find(shape_field + "e2").getKey()
+            else:
+                xxKey = sourceCat.getSchema().find(shape_field + "xx").getKey()
+                yyKey = sourceCat.getSchema().find(shape_field + "yy").getKey()
+                xyKey = sourceCat.getSchema().find(shape_field + "xy").getKey()
+
+            #   If there is a sigma field, get it.  If it also has a signal field
+            #   for signal-to-noise determination, get that too
+            if sigma_field == None:
+                sigmaKey = None
+                sigmaSignalKey = None
+            else:
+                sigmaKey = sourceCat.getSchema().find(sigma_field).getKey()
+                if sigma_signal_field == None:
+                    sigmaSignalKey = None
+                else:
+                    sigmaSignalKey = sourceCat.getSchema().find(sigma_signal_field).getKey()
+
+            #   Now loop through the catalog and summarize the ellipticity measurements
+            for source in sourceCat:
```

Comment:
Instead of looping over sources, it seems like you could just extract column arrays, and do the math with NumPy vectorization.  Should be faster, though I don't know if this is a bottleneck or not.